std/stdio.d

```diff
@@ -1542,6 +1549,185 @@ for every line.
+
+            // Replace the 012345678901234 with random characters
+            rndGen.popFront();
+            auto chars = chain(letters.representation, digits.representation);
+            foreach (ref c; name[prefix.length .. $ - suffix.length])
+            {
+                c = chars[rndGen.front % chars.length];
+                rndGen.popFront();
+            }
+
+            return buildPath(dir, name);
+        }
+
+        // Once should be enough given the randomness, but we'll give it three
+        // times to be safe.
+        enum limit = 3;
+        for (int i = 0; true; ++i)
+        {
+            auto filename = genTempName(prefix, suffix, dir);
+
+            // So that we can test the case where the file already exists.
+            version (unittest)
+            {
+                if (i == 0)
+                {
+                    import std.file;
+                    std.file.write(filename, "nothing");
+                }
+            }
+
+            // O_EXCL does not support symbolic links.
+            if (dirName(filename).isSymlink)
+                filename = buildPath(absolutePath(dirName(filename), baseName(filename)));
+
+            import std.internal.cstring : tempCString;
+            version (Posix)
+            {
+                import core.sys.posix.sys.stat, core.sys.posix.fcntl;
+                auto fd = open(tempCString(filename), O_RDWR | O_CREAT | O_EXCL, S_IRUSR | S_IWUSR);
+            }
+            else version (Windows)
+            {
+                version (DIGITAL_MARS_STDIO)
+                {
+                    enum openFlags = O_RDWR | O_CREAT | O_EXCL;
+                    enum permissions = S_IREAD | S_IWRITE;
+                }
+                else version (MICROSOFT_STDIO)
+                {
+                    enum openFlags = _O_RDWR | _O_CREAT | _O_EXCL;
+                    enum permissions = _S_IREAD | _S_IWRITE;
+                }
+                auto fd = _wsopen(tempCString!wchar(filename), openFlags, _SH_DENYNO, permissions);
+            }
+            else static assert("Unsupported OS");
+
+            if (fd == -1)
+            {
+                import std.string: format;
+                version (unittest)
+                {
+                    import std.file;
+                    if (i == 0 && filename.exists)
+                    {
+                        std.file.remove(filename);
+                        import core.stdc.errno : errno, EEXIST;
+                        assert(errno == EEXIST, format("Error code: %s", errno));
+                    }
+                }
+                if (i == limit - 1)
+                {
+                    auto msg = format("Failed to create temporary file after %s attempts. Last file name tried [%s].",
+                                      limit, filename);
+                    throw new ErrnoException(msg);
+                }
+                continue;
```

Comment:
Would be slightly better to only retry when errno was EEXISTS, i.e. replace the if block above with `errnoEnforce(errno == EEXISTS, msg)`, because none of the other open errors will go away by retying (http://linux.die.net/man/2/open).